### PR TITLE
Automatically detect mainClass

### PR DESCRIPTION
**Pull request by ferozco**
## Before this PR
We required users to provide the main className in the service extension.

## After this PR
==COMMIT_MSG==
Services no longer need to specify the mainClassName, instead it will be inferred from the main source set
==COMMIT_MSG==

## Possible downsides?
N/A

### CODE REVIEW

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LazyCreateStartScriptTask.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Nullable
+    @Override
+    public final String getMainClassName() {
+        return getLazyMainClassName().get();
+    }
+
+    @Override
+    public final void setMainClassName(@Nullable String mainClassName) {
+        throw new UnsupportedOperationException("Use lazy property instead");
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LazyCreateStartScriptTask.java:38]: surely this should just delegate to that `mainClassName.set` ? That way if people had multiple `public static void main(` for some reason, they still have an avenue to override one?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/util/MainClassResolver.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    private static Optional<String> findMainClass(SourceSet sourceSet) {
+        return sourceSet.getAllJava().getFiles().stream()
+                .flatMap(file -> {
+                    String contents = GFileUtils.readFile(file);
+                    if (contents.contains("public static void main(")) {
+                        Matcher matcher = PACKAGE_PATTERN.matcher(contents);
+                        if (matcher.find()) {
+                            return Stream.of(matcher.group(1) + "." + Files.getNameWithoutExtension(file.getName()));
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/util/MainClassResolver.java]: Why are we regexing here? Doesn't the `file` variable that we're flatMapping contain all the path info we need?

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/util/MainClassResolver.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+public final class MainClassResolver {
+    private static final Pattern PACKAGE_PATTERN = Pattern.compile("^package ([^\n;]+)");
+
+    public static String resolveMainClass(Project project) {
+        SourceSet main = project.getConvention().getPlugin(JavaPluginConvention.class)
+                .getSourceSets()
+                .getByName("main");
+        return findMainClass(main).orElseThrow(() -> new GradleException(
+                "Failed to infer main class, please ensure a main class exists"));
```
Comment (iamdanfox) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/util/MainClassResolver.java]: `please ensure you have a method with the signature 'public static void main(String[] args) {}'`

**Comment by iamdanfox** [comment]
I really like this idea!! One more line of gradle bites the dust.  I think @CRogers wanted to do a similar thing in gradle-atlas!

**Comment by ferozco** [comment]
@CRogers was the one with the original idea and it has already been merged and released in gradle-atlas

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LazyCreateStartScriptTask.java (hunk 1 above)
Comment (ferozco) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LazyCreateStartScriptTask.java:38]: if they wanted to override it, they should do so using the extension not by manipulating the task

Comment (iamdanfox) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/tasks/LazyCreateStartScriptTask.java:38]: Ok fair enough

**Review by iamdanfox — APPROVED** [review]

**Comment by iamdanfox** [comment]
While we're on the topic, could we make the default service name just be the project name? So many repos start with:

```gradle

distribution {
    serviceName project.name // <--- always the same!!!
    mainClass 'com.palantir.preview.PreviewServer'
```

**Comment by ferozco** [comment]
Looks like we already do that actually haha

**Comment by iamdanfox** [comment]
🤦‍♂

**Comment by markelliot** [comment]
could you add a test to cover what happens if there are multiple main classes?

**Comment by ferozco** [comment]
@markelliot I've added a test here: https://github.com/palantir/sls-packaging/pull/652

**Pull request merged**